### PR TITLE
Add hcode.  Re-factor rawGet.  Fix infinite loop.

### PR DESCRIPTION
**Pull request by c-blake**
Replace state enum with a cached hash code which has the same memory overhead
and locality as the enum, but can really speed things up with non-integer-like
keys (keys for which either hash() or == take more than couple cycles, or where
the key data is "indirect" and might incur another cache miss).  To function as
both empty/filled state and a hash code cache, it only needs to be ensured that
hash codes are non-zero for any real key.  That is done at the one place in the
whole file hash() is called.  Keep convention clear via isFilled() & isEmpty().
An isDeleted state will no longer be necessary as per below excl/inf loop fix.

Since some use sites know hc and some do not, re-factor rawGet into two forms -
one with known hash code and one with an unknown HC that returns it. Both forms
still return <0 on missing, but returns the much more informative "-1 - index".
That return can be quickly inverted by -1 - result to recover the index where
insert should happen, provided no modifications are made to the table in the
meantime.  This protocol retains the prior <0 interface and also makes it easy
to avoid unnecessary duplicate search work in procs like containsOrInclImpl
(which formerly searched in the initial get and AGAIN in rawInsert).  Strip the
searching part out of rawInsert to "make it even more raw".  swap(s.data, n) a
bit earlier so rawGet and rawGetKnownHC can have similar parameter lists and
integrate well with rawInsert/code sharing between Set and OrderedSet impls.

This PR also fixes infinite looping upon too many deletes. [ The deleted state
(aka "tombstone") approach is vulnerable to the table filling up with deleted
items which forces giant scans for missing keys which could be anywhere.  In
the version prior to this PR, table wraparound wasn't even detected yielding
infinite loops. ] This PR changes excl() from marking slots as deleted to Knuth
algo 6.4R, "local/incremental moveback rehashing" - adapted from Knuth's h->h-1
to the cache-friendlier h->h+1 probe sequence and adapted from "gotos" to a new
doWhile template.  This method restores the table to a state that would have
resulted from pure inserts (in some order).  Update nextTry accordingly.  Since
linear probing can degrade a little faster, 50% rather than 66% may be a better
default growth threshold, but users should be able to adjust threshold anyway.

Old unit tests all pass.  More extensive testing in this module is probably
warranted before taking similar enhancements over to collections.tables.

### CODE REVIEW

**Comment by philip-wernersbach** [comment]
I'd like to see some benchmarks added to the PR before this gets merged. From the forum post there seems to be some hot debate about whether this is appropriate for a variety of workloads. We can't be all things to all people, but we should choose an algorithm that works for a variety of workloads.

**Comment by c-blake** [comment]
# A trivial little program that makes it easy to create pure insert workloads
# with various successful/unsuccessful search mixes at various L1/L2/L3/MainRAM
# scales.  E.g., catenate many small copies to get inputs for mostly successful
# searches or a file w/mostly unique lines for mostly unsuccessful searches.
# You can randomize (or not) order of input lines out-of-band however.
#
# In my timings of this program between old h=5*h+1 and new h=h+1 probes with
# hash caching, the new way always either wins or ties to within measurement
# error..winning by being 1.75X faster for mostly unique files in main memory.
# I haven't tried to "initially right size" the set to distinguish hc speed-up of the
# enlarge vs. regular lookups, but it bears noting that *many* workloads do not
# have initially right-sized tables.

import collections.sets, strutils

when isMainModule:
  proc main() =
    var lineSet = initSet[string]()
    for line in split(readAll(stdin)[0 .. -2], '\l'):
        incl(lineSet, line)
    echo len(lineSet), " distinct lines"
  main()

**Comment by c-blake** [comment]
Yikes.  That sure got formatted crazily, and indents were screwy to boot, but hopefully it makes sense to any but the most casual reader.

**Comment by c-blake** [comment]
Also, that workload doesn't "excl".  It's not very meaningful to do a benchmark with the old version and a good steady-state mix of incl's and excl's because the old one infinite loops.  So, the new one is infinitely faster in interesting cases there. ;-)

Memory utilization and usage are basically the same since I didn't change the growth threshold and just changed an enum to a THash and which I believe are the same size, but correct me if I'm wrong.

A relative evaluation of all vaguely reasonable approaches against many workloads is beyond the scope of my own intent/interest here.

**Comment by c-blake** [comment]
And it's not that I think all vaguely reasonable approaches are bad as mentioned in the forum, but adding lots of table impls to Nim seems a more long-run than short-run goal related to this PR/getting delete working.

As further benchmark color, the 1.75X was whole program time. It''s about 1.85X longer for Nim.old vs Nim.ThisPR if you subtract out the overhead of reading/splitting the file.

If you pre-size the table to its eventual 4 MegaEntries (what I needed for my 120 MB email file in the out of cache test case), then the time ratio drops to only 1.35X faster.   So, (roughly), it's about 1.35X faster to enlarge() with the cached hash code and then h->h+1 is about another 1.35X faster than the old h->5*h+1.  This is in a pure insert workload out of cache, and that as a bonus on top of deletes not blowing up.  Dead entries/tombstones would slow it down further if the workload were mixed.

My input is basically an email file..so, a good deal of entropy per string.  So, other kinds of string lengths and such (or key types!) could easily vary answers as well.  CPUs, RAM speed, compiler options under -d:release, the exact inputs, and yet other things may vary all those numbers.  Everyone has their favorite pet use cases, too. There is no standard workload or corpus of "workload logs" for this that I know of.   If you have a specific benchmark or workload in mind, feel free to describe it.

File: lib/pure/collections/sets.nim [hunk 1]
```diff
@@ -204,6 +245,10 @@ proc incl*[A](s: var HashSet[A], other: HashSet[A]) =
   assert other.isValid, "The set `other` needs to be initialized."
   for item in other: incl(s, item)
 
+template doWhile(a: expr, b: stmt): stmt =
+  b
+  while a: b
```
Comment (Araq) [inline lib/pure/collections/sets.nim]: Your doWhile sux. This one is much better as it avoid the code duplications:

```nimrod

template doWhile(a: expr; b: stmt): stmt = 
  while true:
    b
    if not a: break
```

File: lib/pure/collections/sets.nim [hunk 2]
```diff
@@ -103,25 +114,44 @@ proc mustRehash(length, counter: int): bool {.inline.} =
...
+    # just a few clock cycles, generally worth it for any non-integer-like A.
+    if s.data[h].hcode == hc and s.data[h].key == key:  # compare hc THEN key
       return h
     h = nextTry(h, high(s.data))
-  result = -1
+  result = -1 - h                   # < 0 => MISSING; insert idx = -1 - result
+
+template rawGetImpl() {.dirty.} =
+  hc = hash(key)
+  if hc == 0:       # This almost never taken branch should be very predictable.
```
Comment (Araq) [inline lib/pure/collections/sets.nim]: Why not simply ``hc = hash(key) or 1``?

**Comment by c-blake** [comment]
I addressed your complaint about the doWhile.  Totally fair critique.

"hc = hash(key) or 1" would also work.  A naive analysis might think "they are both 1 CPU instruction" or "branches bad" or "or 1" is 'simpler'.  That was actually my first approach.  Then, in my timings on a SandyBridge i7 with L1 cachable int sets, I found it actually slower by about 10% (1 to 2 cycles, but these are only about 8-10 cycle operations in that context).

I believe this is because the "or 1" form creates a data dependency in the CPU pipeline in the pure integer key case.  On the other hand, hash(key) will almost never equal zero for any sane hash function.  So, if this code is ever run in some inner loop of someone's then the branch predictor will rapidly warm up, the CMP will be 1 cycle, but there will be no data dependency in the "likely taken" branch of non-zero.  So, the CPU speculative execution based on the branch prediction won't have to stall.  It's actually a little tricky to time the fully cold cache/unpredicted case.  That case might be marginally faster with the "or 1" form.  Or they could be the same.  I'd bet on "same" if I had to.

I have a comment partially to this effect.  Some people don't seem to like my detailed explanations.

There's also a vague argument to be made for clarity that the only thing important is that hc not equal zero - not that it's lowest order bit is set.  That is easily addressed by a comment, of course, but some people don't read comments.

Anyway, I don't feel too strongly about "or 1" one way or the other.  If you do feel srongly and don't like my reasons, I am happy to change it.  Or if you prefer "if hc == 0: hc = 1" or some other simpler/less 'cute' value.  Just let me know.

**Comment by c-blake** [comment]
Oh, and I should have said that while I tested it on only a couple generations of Intel core2 CPUs with pure int keys in L1, the explanation probably does generalize to "any pipelined CPU with a branch predictor" - which is to say almost any relevant CPU.  But CPUs vary a lot, too.  I wouldn't be shocked if my comparison varied which is why I don't have strong feelings about it.

Measuring the two two forms "if !=0" vs "or 1" in that cold predictor case is tricky.  It's such a small op that it's tricky even with rdtsc cycle counter timers.  Those rdtsc measurement instructions take order 15-30 cycles on either side and often flush pipelines/etc/etc.  So, the noise in the measurement can be greater than the difference we are trying.  Or you could have an outer loop and do it many times.  That way has to figure out how to trash the predictor tables (which often track hundreds or thousands of code points).  In other words it needs to be very, very outer.  And then there is just so much activity going on..chewing over L1 instruction caches, etc., that again the difference might be lost in the noise.  The "non-innerness" of the loop to have the prediction fail badly probably has to be almost so extreme as to maybe not be very "common case relevant".  The hot cache/hot predictor case really is probably what matters most here.  It's almost certainly what anyone ever doing benchmarking will be looking a because it's hard not to.  In real life, the fully unpredicted case would also probably wind up being in a context of cold L1 instruction caches and just generally be sucking all around in very unusual ways.  Even a 10 cycle difference would probably be nuthin' "in that context".

TL;DR: "my if hc != 0 version is an optimization for the case where the code is being used vaguely often (once in thousands of code branch points) and things can be improved and suckage from everything being cold-cache isn't dominating".

Personally, I'd think my time would be better spent not doing really tricky, subtle benchmarking which are even trickier to make 100% unimpeachable or defending an approach I'm personally sure is best with expansive arguments.  I'd rather try to take these changes over to collections.tables which still infinite loops after all.  At least, that's what I think provided you are not generally against this approach of this PR or wanting to make it contingent on benchmarks that could be argued about forever.

Here's a performance implication hyper summary: "only better or almost unnoticeably worse".

Here are a few more details about the two main modifications (besides fixing the initSize mistake and trying to help users who want to preSize to the right size), there basically being just 3 issues - sequence, hash codes for strings, and hash codes for ints.

Probe sequence: h->h+1 only speeds things up in several cases (steady state insert/delete, lots of inserts of mostly unique keys) and slowed it down in none for me.  The noise at the fastest is tricky.  I tend to do min time of 3 or 10 trials, then do mean/standard deviations over those min times as a nested loop.  So, whatever the workload mix is, LP is very likely to be faster (because any weighted average of the same and faster == faster).  The only bone Jehan had to pick was robustness against sucky hash functions which users can always/should always fix and can always happen with any hashing scheme.  Rather than sacrifice performance for all system libs and users with sane hash()s out of fear of bad hash()s, it seems better to me to help users realize what problem they actually have.  This is actually easy with a debugging mode to track statistically highly unlikely chains and make some noise or something.  Basically just "count loops; then later if loops > 100 and havent_made_noise: make noise", maybe enabled by a debug flag.  Whatever you want, really.

Hash code cache for strings/complex keys: This really speeds up stuff for many situations - e.g. a set of strings with a common prefix like URIs or "fully rooted filesystem paths" where the memcmp()s would have to look at a many bytes..sets/tables growing up from zero which have a lot of enlarge()s, sets with steady state insert/delete that have local re-hashing, etc.  The last delete example would have been true even with isDeleted tombstones (correctly implemented with global rehashes on tombstone pollution) instead of the local rehashing.  The hash code cache only slows down pure int key scenarios and only for successful searches and only by about 2 CPU cycles in my measurements. Even just setting up the loop for a strcmp takes about as long.  It never really hurts and sometimes it really helps.  Jehan even concurred about the hash code, merely trying to temper expectations.  It makes the worst case he was concerned about boil down to code almost identical to a "non-MMX/SSE-optimized strlen".  The scan is just for a matching hcode instead of zero and has a larger stride than 1 byte.  The h->h+1 interacts pleasantly with that - it becomes a scan over local memory not hopping all over the table.  It also takes up no more space than the old enum did.  Indeed, the only valid arguments against a hash code IMO are A) space and B) slows ints.  The space was already spent in the old version, but used only 2 values out of 2**32 or 2**64.

The hash code cache cost for ints can and should be avoided someday with the obvious specializations, already mentioned in my small comment in the intial PR: one with no special value and no hash code and another with 1 special empty value which could also use 1/2 the RAM or maybe even less (1/4 the RAM for 4-byte ints).  The only open questions to me here are programming language/style issues.  Could/should those specializations share names/the interface/etc. with the main Set/Table.  They have both different representations (no .hcode) and different interfaces (the need for the user to specify the value signifying empty).  That seems more like a long-term open question.  In the short-term and in a broader context, 2 cycles is much, much less than a function call (about 14 on my system).  Only truly extreme int-keyed users will ever notice any difference one way or the other.  Truly, the 1/2 the RAM is likely a more significant optimization, and that requires eliminating the enum and the hash code.  Bit vectors are better for a lot of extreme int-key stuff, but not all.  E.g., iterating over a large, sparse set can be slow.  A truly specialized int-keyed set (but not table) with special empty values for smaller integer sizes could even be SSE optimized just like strlen, maybe even by a good auto-vectorizer if the "mod tablesize" wraparound handling didn't kill the benefit.  Anyway, that really is a special case.

Anyway, I'm just trying to help out here.  Nim doesn't tie syntax to these libs in any strong way (a virtue!).  I can always have all my own stuff privately, replace stdlibs local dirs, etc.  That's how I was working initially.  I thought I would try to give back.  A lot of people seem to check out Nim with some simplecookie-cutter table calculations and get impressed.  I think they could be impressed even more.  You can trust me on the relative timings stuff or run your own.

**Comment by c-blake** [comment]
Another way in which h->h+1 addresses a different than mis-use "worst case scenario" is just the worst case for hashing in general - namely the largest/highest latency memory/table possible.  For those you can really measure the improvement of staying local vs. hopping around gigabytes to avoid clustering, and that diff is much more than "3%" bandied about on IRC in relation to Haskell guidelines.  So, the probe sequence also falls under the "is the same or a bit faster most of the time and much better when you most need it" argument, from a certain point of view.

**Comment by c-blake** [comment]
Also on the "or 1" question, there are subtle points I glossed over before that may bear mentioning.

Though I referred to it as "or 1" in shorthand to back reference your text, what I actually timed against "if hc == 0" was "hash(key) or 1073741824" (aka 2**30), not vs "or 1".  My high-bit-or version preserves initial probe diversity with "and table_size" reduction.  Ok, ok, I actually had in my code was a "when:" in there for the 32/64 bit systems, too, since a 64-bit system might have many giga-entries.  Being only 2**30 was to just simplify life for the purposes of that test..there might be some trivial workaround to set the highest order bit instead, but I was testing with 10s of mega-entry not giga-entry tables.  I seriously doubt the timing results/effects would differ between 1st and 2nd highest bit.  So, "if hc == 0" is also more portable (admittedly implementing that portability has no run-time cost, but it looks uglier/more complex than the literal "or 1" alternative you mention).

Literally unconditionally turning on the lowest order bit as in your "why not or 1" question is not what I tried.  That makes *every hash be an odd number*.  With a range reduction to table space of "and table_size", this is bad.  It's almost like having a table half as large or a table load factor twice as big (though only for initial probes).  You can restore randomness in the hash output by shifting off said mandatory low bit, but then you need to do said transform *after* you've used it's non-zero-ness to decide emptiness.  That later transforming creates more work in probing, etc., not just the lesser work at hash(x) time.  Even C bit fields that had to "and"/"shift" on accesses would be decidedly slower than the as-is-friendly "if hc == 0" approach that I went with which only excludes a single value from being a valid hash, not 1/2 the values.  A slightly larger comment/explanation in the code about this may be warranted.  I can add that comment if you want.  This seemed a maybe non-obvious enough gotcha to be worth following up on.

If, by some chance, you really hate "if hc == 0" and had your heart set on "or low_bit" instead of "or high_bit", you could still get "as-is" .hcode's at about the same cost if range reduction to table space was via a "shr (32 - bits)" or shr (64 - bits).  *However*, that approach would interact with existing hash functions (like the identity for ints).  Instead of nearby inputs (whatever +-1, +-2 etc.) hashing to distinct slots, you would instead get "nearby modulo table_size" mapping to distinct spots.  While not un-doable, it would tend to make one reevaluate all the standard hashes.nim procs for their collision properties (in regard to their high rather than low bits).  "if hc == 0" as the current PR has seems cleaner to me, and as I mentioned ever so slightly more CPU pipeline friendly.

**Comment by Araq** [comment]
The PR is fine, will merge it later, but please show a *little* more respect for my time. There is no way I'm gonna read all of this.

**Comment by c-blake** [comment]
I will try to keep it shorter.  Length is never intended to disrespect anything. :) The sets.nim stuff seems good to merge to me whenever you want.  I'm moving onto tables.nim as per my plan.

I did want to raise one question with just two paragraphs of background..Doing dup keys in an open addressed table is unusual and of limited utility..It's hard to know what del will even delete.  enlarge() breaks any kind of FIFO discipline at the high index end of the table (iterating over wrapped around keys before pre-end keys when creating the enlarged table).  Granted, table order is already unreliable..and maybe del could return the deleted value.  Performance can also crash if dups in the table are "not rare".  Also, the table type doesn't distinguish tables with dups from those with only unique keys - a user must know the history of an object to know if they must use allValues or if they must loop - del()ing until get fails to really get rid of a key.

Do you know of anyone even using this dup key feature?  I couldn't find allValues() used anywhere.  The only use case I found in a cursory search of the Nim repository is in lib/pure/fsmonitor.nim.  That use is arguably "confused".  I am pretty sure it does not need the extra flexibility of 'add' over 'put'.  I'd bet Dominik just thought of 'add' was doing what 'put' does (unless back when it was written they were the same).  They are similar words.  Though there is no allValues or special dup possibility handling that I see, fsmonitor probably works fine since the keys in that module (watch descriptors & event cookies) are likely kept unique by construction of the OS.  My bet would be that few Nim users are even aware of the feature.

So, I _think_ I can keep the dup key stuff working, but it would be an easier re-factor if I did not, and I am thinking maybe I should not keep it.  My question to you is this: Are you certain the pros of providing add() outweigh the cons?  Do you really want to keep it?  Table[A, seq[B]] is also pretty easy to do.  Most programming environments would at least have a separate table type or flag inside objects to distinguish tables that can have dups vs. those with unique keys.  It's also probably important to settle that interface question before 1.0.

**Comment by Araq** [comment]
Well I use tables like these in the compiler where it's an essential feature (overloaded symbols). I prefer it over 'Table[A, seq[B]]' because IMHO 'Table[A, seq[B]]' means that you didn't understand what hash tables are capable of. ;-) On the contrary, I consider 'del' the feature that is just not necessary in practice and that makes you pay a price for nothing.

In any way, if you want to get rid of it, you better provide some kind of deprecation path.

**Comment by c-blake** [comment]
Ok.  Just making sure it was actively wanted.  I'll try to keep it.

**Pull request merged**